packages/cli/src/cmds/agentInstaller/projectConfiguration.ts

```diff
@@ -81,12 +87,109 @@ async function resolveSelectedInstallers(
     }
   }
 
-  // Sanity check. Verify everything resolved properly. This should branch should never occur.
+  // Sanity check. Verify everything resolved properly. This branch should never occur.
   if (projects.some(({ selectedInstaller }) => !selectedInstaller)) {
     throw new Error('Invalid selection');
   }
 }
 
+type Yarn1Workspace = {
+  location: string;
+};
+
+// With yarn 1:
+// - if there's no "workspaces" key in package.json we get the following error:
+//   $ yarn workspaces info --json
+//   yarn workspaces v1.22.19
+//   error Cannot find the root of your workspace - are you sure you're currently in a workspace?
+//   info Visit https://yarnpkg.com/en/docs/cli/workspaces for documentation about this command.
+// - if there is a "workspaces" key in package.json the output contains text
+//   outside of json. Remove it:
+//   $ yarn workspaces info --json
+//   yarn workspaces v1.22.19
+//   {
+//    "workspace-a": {
+//      "location": "workspace-a",
+//      "workspaceDependencies": [],
+//      "mismatchedWorkspaceDependencies": []
+//    }
+//   }
+//   Done in 0.02s.
+async function getYarnSubprojectsVersionOne(
+  dir: string,
+  installer: YarnInstaller,
+  subprojects: ProjectConfiguration[]
+) {
+  const cmd = new CommandStruct('yarn', ['workspaces', 'info', '--json'], installer.path);
+  const output = await run(cmd);
+
+  // Remove any text before and after JSON
+  let lines = output.stdout.split('\n');
+  let outputClean = '';
+  for (const line of lines) {
+    if (line.startsWith('error')) return;
+    if (!(line.startsWith('yarn workspaces v') || line.startsWith('Done in '))) outputClean += line;
+  }
+
+  Object.entries(JSON.parse(outputClean)).forEach(([workspaceName, workspaceData]) => {
+    const data = workspaceData as Yarn1Workspace; // is there a better way to use this type?
```

Comment:
A better way would be to verify it's the expected type and let TypeScript know about it. (Since the type is very small and only used here, it might not be necessary to even declare it.) Here's a suggestion:
```typescript
function yarn1Workspaces(info: unknown): [string, string][] {
  assert(info && typeof info === 'object');
  return Object.entries(info).map(([name, data]: [string, unknown]) => {
    assert(data && typeof data === 'object');
    assert('location' in data);
    assert(typeof data.location === 'string');
    return [name, data.location];
  });
}

for (const [name, location] of yarn1Workspaces(JSON.parse(outputClean))) {
  const path = join(dir, location),
  subprojects.push({
    name: basename(name), path
    availableInstallers: [new YarnInstaller(path)]
  });
}
```
Note how `unknown`s force the type checker to make sure you're verifying the object types before using them.

Reply:
Changed.